locator/network_topology_strategy.cc

```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+}
+
+future<tablet_map> network_topology_strategy::reallocate_tablets(schema_ptr s, token_metadata_ptr tm, size_t tablet_count, const tablet_map* cur_tablets) const {
+    static thread_local std::default_random_engine rnd_engine{std::random_device{}()};
+
+    natural_endpoints_tracker::check_enough_endpoints(*tm, _dc_rep_factor);
+
+    auto tablets = cur_tablets ? *cur_tablets : tablet_map(tablet_count);
     load_sketch load(tm);
     co_await load.populate();
 
-    // FIXME: Don't use tokens to distribute nodes.
-    // The following reuses the existing token-based algorithm used by NetworkTopologyStrategy.
-    assert(!tm->sorted_tokens().empty());
-    auto token_range = tm->ring_range(dht::token::get_random_token());
-
+    const auto& topo = tm->get_topology();
+    const auto dc_rack_nodes = topo.get_datacenter_rack_nodes();
     for (tablet_id tb : tablets.tablet_ids()) {
-        natural_endpoints_tracker tracker(*tm, _dc_rep_factor);
-
-        while (true) {
-            co_await coroutine::maybe_yield();
-            if (token_range.begin() == token_range.end()) {
-                token_range = tm->ring_range(dht::minimum_token());
+        tablet_replica_set replicas;
+        struct rack_load {
+            sstring rack_name;
+            size_t load;
+        };
+        struct rack_load_cmp {
+            bool operator()(const rack_load& a, const rack_load& b) const {
+                return a.load > b.load;
             }
-            locator::host_id ep = *tm->get_endpoint(*token_range.begin());
-            token_range.drop_front();
-            if (tracker.add_endpoint_and_check_if_done(ep)) {
-                break;
+        };
+        struct dc_rack_load {
+            size_t node_count = 0;
+            std::vector<rack_load> racks;
+        };
+        // Current load per dc/rack
+        std::unordered_map<sstring, dc_rack_load> dc_rack_load_map;
+        // Current replicas per dc/rack
+        std::unordered_map<sstring, std::unordered_map<sstring, std::unordered_set<locator::host_id>>> replicas_per_dc;
+        if (cur_tablets) {
+            replicas = cur_tablets->get_tablet_info(tb).replicas;
+            for (const auto& tr : replicas) {
+                const auto& node = topo.get_node(tr.host);
+                replicas_per_dc[node.dc_rack().dc][node.dc_rack().rack].insert(tr.host);
+                ++dc_rack_load_map[node.dc_rack().dc].node_count;
+            }
+            for (const auto& [dc, racks] : replicas_per_dc) {
+                auto& dr_load = dc_rack_load_map[dc];
+                dr_load.racks = boost::copy_range<std::vector<rack_load>>(racks | boost::adaptors::transformed([] (const auto& x) {
+                    return rack_load(x.first, x.second.size());
+                }));
             }
+        } else {
+            replicas.reserve(_rep_factor);
         }
+        for (const auto& [dc, dc_rf] : _dc_rep_factor) {
+            auto& dr_load = dc_rack_load_map[dc];
```

Comment:
For the sake of readability improvement -- does it make sense to wrap the hole body of this loop into reallocate_tablets_in_dc()?

Reply:
yes, we can refactor the loop body out into an internal function.

Reply:
refactored in v2